### PR TITLE
LuGre friction model implementation

### PR DESCRIPTION
**Pull request by ale-git**
The LuGre friction model has been implemented in SImulink, ready for automatic code generation.
The output code is included in icub-firmware to add sensorless torque estimation capabilities to the control boards.

### CODE REVIEW

**Comment by pattacini** [comment]
Thanks @ale-git for the contribution 👍🏻 
I'll be reviewing the PR in the next few days.

**Comment by pattacini** [comment]
Hi @ale-git 

What's the strategy behind parsing the LuGre params?
Ideally, to provide back-compatibility (quite important here), if the section is not found then the SW should NOT make use of the LuGre.

**Review by pattacini — CHANGES_REQUESTED** [review]
Hi @ale-git 

Before merging this PR, we need to address several points.

### Documentation
- The state equations is not correct.
- Missing details on $`g\left(v\right)`$.
- It'd be appropriate to cite a paper on LuGre, e.g. DOI: 10.1109/mcs.2008.929425.
- Nothing is said about how to tune the LuGre params. This is fundamental to make it usable.

### Implementation
- Especially when a dynamic is at stake, it's much better to resort to Simulink rather than MATLAB code. We already have an internal LuGre implementation in Simulink that I can refer you to[^1].
- Most of the time $`\sigma_1`$ is related to $`\sigma_0`$.
- The model interface can be improved.

### MATLAB Project
The project has to be fine-tuned, although this is the very last thing.

---

I propose to set up a meeting for this 🗓️

[^1]: It looks like this:

      image

**Comment by ale-git** [comment]
I commited the wrong README.md file, whose right version was deleted instead (fixed).
I also added a couple figures.

**Comment by pattacini** [comment]
Any update on this?

**Comment by ale-git** [comment]
I've added the Simulink-block made torque estimator to the model.
image

**Comment by pattacini** [comment]
Great!
There were other things to address we discussed last week.
I'll give the PR the last touch when the work is complete.

**Comment by pattacini** [comment]
Hi @ale-git 

We still need to address the following important point:

>Nothing is said about how to tune the LuGre params. This is fundamental to make it usable.

File: sensorless_torque/README.md [hunk 1]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
 that corresponds to the folder `icub-firmware\emBODY\eBcode\arch-arm\mbd\torque_estimator`
 
 ![folder](assets/Folder.png)
+
+
+# LuGre parameters estimation
+An automatic parameter estimation tool is provided in this folder:
+
+[param_estimator.m](param_estimator.m)
```
Comment (pattacini) [inline sensorless_torque/README.md:123]:
```suggestion
An automatic parameter estimation tool is provided:
- [`param_estimator.m`](param_estimator.m)
```

File: sensorless_torque/README.md [hunk 2]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
...
+
+The data source experiment should be in some way meaningful, i.e. with typical velocity and torque back and forth motion cycles, duration > 20 s, about 1 kHz sampling. 
+
+Example data file: [ergojoint_mid_calibration.mat](data/ergojoint_mid_calibration.mat)
+
+![Example data plot](assets/DataExample.png)
+
+The file path must be inserted in the script first lines, here:
+
+![Input filepath](assets/filepath.png)
```
Comment (pattacini) [inline sensorless_torque/README.md:151]:
```suggestion
![Input filepath](assets/FilePath.png)
```

File: sensorless_torque/README.md [hunk 3]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
...
+An automatic parameter estimation tool is provided in this folder:
+
+[param_estimator.m](param_estimator.m)
+
+The algorithm uses brute force in iterative steps, starting from a wide range of possible values and coarse quantization, and going on narrowing the range and increasing the resolution at each step.  
+Since the number of parameters is still quite high, the algorithm estimates the LuGre parameters in 3 stages:
+
+1.  Km (motor torque constant), Kw (viscous friction constant), Fc (Coulomb force constant)
+2.  Fs (Stribeck force constant), Vth (Stribeck force velocity threshold)
+3.  S0, S1 (LuGre hysteresis model stiffness and damping).
```
Comment (pattacini) [inline sensorless_torque/README.md:130]:
```suggestion
1.  `Km` (motor torque constant), `Kw` (viscous friction constant), `Fc` (Coulomb force constant);
2.  `Fs` (Stribeck force constant), `Vth` (Stribeck force velocity threshold);
3.  `S0`, `S1` (LuGre hysteresis model stiffness and damping).
```

This formatting helps highlight variables within the text.

File: sensorless_torque/README.md [hunk 4]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
...
+
+The search space must be initially limited in some way, thus the Km torque constant, Kw viscous friction constant and Fc Coulomb force constant initial ranges are restricted to 1/5 to 5x the values measured in the ErgoCub new medium joint, covering in this way a wide range of present and future joint sizes.
```
Comment (pattacini) [inline sensorless_torque/README.md:132]:
```suggestion
The search space must be initially limited in some way, thus the `Km` torque constant, `Kw` viscous friction constant and `Fc` Coulomb force constant initial ranges are restricted to 1/5x to 5x the values measured in the ergoCub new medium joint, covering in this way a wide range of present and future joint sizes.
```

File: sensorless_torque/README.md [hunk 5]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
...
+
+
+## Data input format
+The script requires an input with a 4xN matrix containing:
+
+1.  time [s]
+2.  measured current [A]
+3.  measured velocity [rad/s], motor side
+4.  measured joint torque [Nm], joint side.
```
Comment (pattacini) [inline sensorless_torque/README.md:141]:
```suggestion
1.  time [s];
2.  measured current [A];
3.  measured velocity [rad/s], motor side;
4.  measured joint torque [Nm], joint side.
```

File: sensorless_torque/README.md [hunk 6]
```diff
@@ -113,3 +115,44 @@ For example, with the EMS board:
...
+
+The data source experiment should be in some way meaningful, i.e. with typical velocity and torque back and forth motion cycles, duration > 20 s, about 1 kHz sampling.
```
Comment (pattacini) [inline sensorless_torque/README.md:143]:
```suggestion
The data source experiment should be in some way meaningful: i.e., undergoing typical velocity and torque, back and forth motion cycles, time duration > 20 s, about 1 kHz sample rate. 
```

File: sensorless_torque/assets/Output.png
Comment (pattacini) [inline sensorless_torque/assets/Output.png:1]: I think it's important to output the measurements units too.

**Review by pattacini — CHANGES_REQUESTED** [review]
Hi @ale-git 

Thanks for the notes on how to tune the params.
I've dropped some inline suggestions you may incorporate straight away.

A couple of important points though:
- Nothing is said about the math underlying the brute force estimation. I think that we ought to outline the principle beneath at least, in a brief paragraph.
- I would report the measurement units along with the values.

**Comment by ale-git** [comment]
I've made the requested changes

**Review by pattacini — APPROVED** [review]
Thanks!

**Pull request merged**